File: 101. Symmetric Tree/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        def s(a, b):
            if not a and not b:
                return True
            if not a or not b:
                return False
            return a.val == b.val and s(a.left, b.right) and s(a.right, b.left)
        return s(root.left, root.right)

File: 101. Symmetric Tree/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        //helper function to check if two subtrees are mirrors
        function<bool(TreeNode*, TreeNode*)> s = [&](TreeNode* a, TreeNode* b) -> bool {
            if (!a && !b) return true;
            if (!a || !b) return false;
            return (a->val == b->val) && 
                   s(a->left, b->right) && 
                   s(a->right, b->left);
        };
        
        return s(root->left, root->right);
    }
};
